File: backend/routers/watchlists.py
from fastapi import APIRouter
from pydantic import BaseModel
from services.watchlist_sheets import (
    read_watchlists, save_watchlist, delete_watchlist,
)

router = APIRouter()


class WatchlistBody(BaseModel):
    filter: dict = {}


@router.get("/watchlists")
async def get_watchlists():
    try:
        results = await read_watchlists()
        return {"results": [w.model_dump() for w in results]}
    except Exception as e:
        return {"error": str(e), "results": []}


@router.put("/watchlists/{name}")
async def put_watchlist(name: str, body: WatchlistBody):
    n = name.strip()
    if not n:
        return {"saved": False, "error": "Watchlist name is required"}
    try:
        w = await save_watchlist(n, body.filter)
        return {"saved": True, **w.model_dump()}
    except Exception as e:
        return {"saved": False, "error": str(e)}


@router.delete("/watchlists/{name}")
async def remove_watchlist(name: str):
    n = name.strip()
    try:
        removed = await delete_watchlist(n)
        if not removed:
            return {"deleted": False, "error": f"No watchlist named {n!r}"}
        return {"deleted": True, "name": n}
    except Exception as e:
        return {"deleted": False, "error": str(e)}

File: frontend/src/lib/watchlists.ts
import { API_BASE as API } from './api'

export type NumRange = { min: number | null; max: number | null }

export interface SerializedFilters {
  tickers: string[]
  stockTypes: string[]
  quality: NumRange
  gap: NumRange
}

export interface Watchlist {
  name: string
  filter: SerializedFilters
  created: string
}

export async function fetchWatchlists(): Promise<Watchlist[]> {
  const res = await fetch(`${API}/api/watchlists`)
  const data = await res.json()
  if (data.error) throw new Error(data.error)
  return (data.results ?? []) as Watchlist[]
}

export async function saveWatchlist(name: string, filter: SerializedFilters): Promise<void> {
  const res = await fetch(`${API}/api/watchlists/${encodeURIComponent(name)}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ filter }),
  })
  const data = await res.json()
  if (!data.saved) throw new Error(data.error || 'Failed to save watchlist')
}

export async function deleteWatchlist(name: string): Promise<void> {
  const res = await fetch(`${API}/api/watchlists/${encodeURIComponent(name)}`, {
    method: 'DELETE',
  })
  const data = await res.json()
  if (!data.deleted) throw new Error(data.error || 'Failed to delete watchlist')
}
